### PR TITLE
whatmp3 3.6

### PR DESCRIPTION
**Pull request by guillaumekh**
Hi, this is both my very first pull request and my first Homebrew formula. I tried to follow all guidelines & documentation. I hope it all went OK :pray: 

Whatmp3 is a "Python script to convert FLAC to to mp3/ogg/alac/aac and create a new torrent for you with mktorrent."
It has been maintained for +6 years now.
It's already on AUR https://aur.archlinux.org/packages/whatmp3
I'm not the author/maintainer.

### CODE REVIEW

File: Library/Formula/whatmp3.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  depends_on "vorbisgain" => :optional
+  depends_on "sox" => :optional
+
+  def install
+    bin.install "whatmp3"
+  end
+
+  test do
+    output = shell_output("#{bin}/whatmp3 --version")
+    assert_match /whatmp3 3.6/, output
```
Comment (bfontaine) [inline Library/Formula/whatmp3.rb]: We need a test that does something more substantial than e.g. `--version`/`--help`. See the [`flac.rb`](https://github.com/Homebrew/homebrew/blob/31db1e66c71a2378c847070d58f143af4f39b068/Library/Formula/flac.rb#L68) for a test with a dummy FLAC file. `#{bin}/whatmp3 --320 -n something.flac` will convert a file but not create a torrent so we could use that. Please ask if you have any question.

**Comment by guillaumekh** [comment]
OK, i think i got it (thanks for the helpful feedback).

**Comment by guillaumekh** [comment]
This doesn't work. That dummy file cannot be decoded since it's not a valid FLAC file.

```bash
$ brew test whatmp3 --verbose
Testing whatmp3
==> /usr/local/Cellar/whatmp3/3.6/bin/whatmp3 --notorrent --V0 flac
/private/tmp/whatmp320151015-27364-1bi0h5i/flac/file.flac: ERROR: reading metadata, status = "FLAC__METADATA_CHAIN_STATUS_NOT_A_FLAC_FILE"
[...]
The file does not appear to be a FLAC file.
[...]
file.flac: *** Got error code 0:FLAC__STREAM_DECODER_ERROR_STATUS_LOST_SYNC

The input file is either not a FLAC file or is corrupted.  If you are
convinced it is a FLAC file, you can rerun the same command and add the
-F parameter to try and recover as much as possible from the file.


file.flac: ERROR while decoding metadata
           state = FLAC__STREAM_DECODER_END_OF_STREAM
Warning: unsupported audio format
Encoding with V0 started...

Encoding with V0 finished.
Error: whatmp3: failed
<false> is not true.
```

I figured since FLAC is a dependency it was OK-ish to call it.
Anyway, I've made a smallest-possible FLAC file from scratch and converted it to string. The flac dependency is removed.

File: Library/Formula/whatmp3.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  depends_on "vorbisgain" => :optional
+  depends_on "sox" => :optional
+
+  def install
+    bin.install "whatmp3"
+  end
+
+  test do
+    # Create dummy FLAC
+    Dir.mkdir("flac")
```
Comment (bfontaine) [inline Library/Formula/whatmp3.rb]: Use `(testpath/"flac").mkpath`.

File: Library/Formula/whatmp3.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    (testpath/"flac/file.flac").write "fLaC\x00\x00\x00\"\x04\x80\x04\x80\x00\x00\f\x00\x00\f\x01\xF4\x00\xF0\x00\x00\x00\x01\xF3\x8B\xE3\xDBM\x93\xE40\\~$\xBE\x94\xEF\x01\x9A\x84\x00\x00( \x00\x00\x00reference libFLAC 1.2.1 20070917\x00\x00\x00\x00\xFF\xF8d\b\x00\x00\xE3\x03\x01\xFD\xEC\x10"
+    # Decode with Whatmp3
```
Comment (bfontaine) [inline Library/Formula/whatmp3.rb]: You can remove this comment.

**Comment by guillaumekh** [comment]
I also noticed the Travis build fails because of a python/permission issue : 
```bash
==> brew test --verbose whatmp3
Testing whatmp3
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20151015-7796-a0t79r.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Library/Homebrew -- /usr/local/Library/Homebrew/test.rb /usr/local/Library/Formula/whatmp3.rb --verbose
==> /usr/local/Cellar/whatmp3/3.6/bin/whatmp3 --notorrent --V0 flac
.
==> Sandbox log
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/optparse.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gettext.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/__init__.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/util.pyc
Error: whatmp3: failed
Failed assertion, no message given.
==> FAILED
```

These .pyc files are Core Python modules, which whatmp3 imports on launch. They should not be posing any problem. 
I have tried replacing the install method with the one documented in the [Homebrew wiki for Python applications](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Python-for-Formula-Authors.md), which gives the following code : 
```ruby
def install
  ENV.prepend_create_path "PYTHONPATH", libexec/"lib/python2.7/site-packages"
  system "python", *Language::Python.setup_install_args(libexec)

  bin.install Dir[libexec/"bin/*"]
  bin.env_script_all_files(libexec/"bin", :PYTHONPATH => ENV["PYTHONPATH"])
end
```

However that install method fails with the following error :
```bash
Last 15 lines from /Users/guillaumekhayat/Library/Logs/Homebrew/whatmp3/01.python:
python
-c
import setuptools, tokenize
__file__ = 'setup.py'
exec(compile(getattr(tokenize, 'open', open)(__file__).read()
  .replace('\r\n', '\n'), __file__, 'exec'))
--no-user-cfg
install
--prefix=/usr/local/Cellar/whatmp3/3.6/libexec
--single-version-externally-managed
--record=installed.txt

Traceback (most recent call last):
  File "<string>", line 3, in <module>
IOError: [Errno 2] No such file or directory: 'setup.py'
```

Do you have a clue what's going on ? Thanks for taking the time to assist.

**Comment by bfontaine** [comment]
```
==> Sandbox log
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/optparse.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gettext.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/__init__.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.pyc
Oct 15 22:44:59 sandboxd[124]: Python(7807) deny file-write-create /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/util.pyc
```

The sandbox log says that it tried to write `.pyc` files in another formula prefix; which is forbidden by our sandbox rules. I guess it uses these files and Python tries to save the precompiled versions.
@tdsmith Any idea how we can prevent this?

**Comment by guillaumekh** [comment]
Found that from ["What’s New in Python 2.6 - Interpreter Changes"](https://docs.python.org/2/using/cmdline.html#envvar-PYTHONDONTWRITEBYTECODE):
> Python can now be prevented from writing .pyc or .pyo files by supplying the -B switch to the Python interpreter, or by setting the PYTHONDONTWRITEBYTECODE environment variable before running the interpreter. This setting is available to Python programs as the sys.dont_write_bytecode variable, and Python code can change the value to modify the interpreter’s behaviour.

I'm not sure exactly how to set that env var properly in the test method.

**Comment by bfontaine** [comment]
@guillaumekh Try `ENV["PYTHONDONTWRITEBYTECODE"] = "1"`.

**Comment by tdsmith** [comment]
The .pyc files shouldn't cause trouble and failure to write them doesn't cause errors.

**Comment by tdsmith** [comment]
(The test is failing for other reasons.)

**Comment by guillaumekh** [comment]
OK, so it looks like taking the high road here would mean leaving bytecode/.pyc generation enabled. However, without more verbose output from the CI machine, I'm unsure how to fix this.

A good place to start seems with the install method. I'm using ```bin.install "whatmp3"``` but that [seems inappropriate](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Python-for-Formula-Authors.md), can you please first confirm we should be installing with this method instead :
```Ruby
def install
  ENV.prepend_create_path "PYTHONPATH", libexec/"lib/python2.7/site-packages"
  system "python", *Language::Python.setup_install_args(libexec)

  bin.install Dir[libexec/"bin/*"]
  bin.env_script_all_files(libexec/"bin", :PYTHONPATH => ENV["PYTHONPATH"])
end
```

**Comment by bfontaine** [comment]
The test is failing:
```
==> /usr/local/Cellar/whatmp3/3.6/bin/whatmp3 --notorrent --V0 flac
sh: lame: command not found

flac 1.3.1, Copyright (C) 2000-2009  Josh Coalson, 2011-2014  Xiph.Org Foundation
flac comes with ABSOLUTELY NO WARRANTY.  This is free software, and you are
welcome to redistribute it under certain conditions.  Type `flac' for details.

file.flac: done         
Encoding with V0 started...

Encoding with V0 finished.
Error: whatmp3: failed
```

It seems that `lame` is required for this test.

**Comment by guillaumekh** [comment]
Switched lame dependency from ```:optional``` to ```:recommended``` which should match most users' needs anyway.
Homebrew doesn't support a ```:test``` flag for dependency, does it ? Couldn't find any such thing in the Formula Cookbook.

**Comment by guillaumekh** [comment]
The Mavericks & Yosemite Jenkins builds passed but the El Capitan build failed with error ```Slave went offline during the build```

This looks like an Jenkins issue rather than a whatmp3 issue. @bfontaine can you trigger a new build ?

**Comment by mikemcquaid** [comment]
@BrewTestBot test this please

**Comment by bfontaine** [comment]
> Homebrew doesn't support a `:test` flag for dependency, does it ?

Nope.

**Comment by guillaumekh** [comment]
Unless I'm missing something, it looks like we're good. Please let me know if i need to make further changes.

File: Library/Formula/whatmp3.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+class Whatmp3 < Formula
+  desc "whatmp3 is a small script to create mp3 torrents out of FLACs."
```
Comment (bfontaine) [inline Library/Formula/whatmp3.rb]: You can shorten this into “`Small script to create mp3 torrents out of FLACs`”.

**Comment by guillaumekh** [comment]
Done. Description changed as requested

**Comment by guillaumekh** [comment]
Hey, didn't want to be pushy but unless I'm missing something, it looks like we're good. Please let me know if i need to make further changes.

**Comment by bfontaine** [comment]
Thank you for your contribution to Homebrew @guillaumekh; we appreciate it! :smile:

**Pull request closed without merging**